_posts/2019-05-20-server-rendering-responsively.md

```diff
@@ -0,0 +1,201 @@
+
+Yuck! The docs explain that `ReactDOM.hydrate` doesn't put a lot of effort into patching up differences between the
+server and client renders, and that's how you end up with weird things like above. Weird things don't _always_
+happen, but they _can_ happen. Your best bet is to match server and client renders, to avoid strange results that
+are difficult to explain or debug.
+
+Because hydration requires the component trees to match between server and client, we render _all_ breakpoints from
+the server with `react-responsive-media`. This way, we know you're going to get a branch that matches between the
+client and server. Then, upon hydration, we prune the branches that don't match the current breakpoint.
+
+If you're interested, you can
+[read the issue that originally inspired us to build `@artsy/react-responsive-media`](https://github.com/artsy/reaction/issues/1367).
+One of the neat things about Artsy being [open-source by default](/series/open-source-by-default/) is that you can
+see decisions being made and libraries being built _as they happen_; not just after they're complete.
+
+## Tool 3: [`@artsy/detect-responsive-traits`](https://github.com/artsy/detect-responsive-traits)
+
+I mentioned above that it's difficult to accurately detect devices by user agent to identify which breakpoint to
+render. We didn't want this to be our primary strategy for combining SSR with responsive design.
+
+But with `react-responsive-media` as our primary approach, we felt that we could make some further optimizations
+with user agent detection. In the event that we don't know your device by its user agent, we'll still give you both
+the desktop and mobile breakpoints from SSR. But if we are certain you are on a mobile device, we can omit the
+desktop breakpoint from the server render.
+
+We really wanted to not maintain our own list of user agents. Alas, we found that none of the existing user agent
+detection libraries surfaced all the information we needed. We needed to know the minimum width for a browser on a
+given device, and if it was resizable, and to what dimensions it was resizable. If any existing libraries _did_
+have this data, they didn't provide it to us easily.
+
+So we did some experimentation, given the browsers and devices we knew we needed to support. And yeah...we
+(reluctantly) created our own user-agent detection library,
+[@artsy/detect-responsive-traits](https://github.com/artsy/detect-responsive-traits). We're using this to determine
+if your browser is likely going to use only the mobile breakpoint of our app, in which case we don't have to also
+render the desktop version. The library is currently targeting only the browsers and devices we support on
+[artsy.net](artsy.net), but
+[we're always open to contributions](https://github.com/artsy/detect-responsive-traits)!
+
+We aren't doing any detection of desktop browsers. Maybe this is something we do in the future, but for now we are
```

Comment:
```suggestion
We aren't doing any detection of desktop browsers, as they are generally resizable to any size. For now we are
```

You may want to rework that second sentence, but I just want to make clear that there isn’t any point in sniffing desktop user agents.